src/Servers/Kestrel/Core/src/Internal/Http/HttpProtocol.cs

```diff
@@ -733,108 +743,55 @@ public void OnCompleted(Func<object, Task> callback, object state)
         protected Task FireOnStarting()
         {
             var onStarting = _onStarting;
-
-            if (onStarting == null || onStarting.Count == 0)
-            {
-                return Task.CompletedTask;
-            }
-            else
+            if (onStarting?.Count > 0)
             {
-                return FireOnStartingMayAwait(onStarting);
+                return ProcessEvents(this, onStarting);
             }
-        }
 
-        private Task FireOnStartingMayAwait(Stack<KeyValuePair<Func<object, Task>, object>> onStarting)
-        {
-            try
+            return Task.CompletedTask;
+
+            static async Task ProcessEvents(HttpProtocol protocol, Stack<KeyValuePair<Func<object, Task>, object>> events)
             {
-                while (onStarting.TryPop(out var entry))
+                // Try/Catch is outside the loop as any error that occurs is before the request starts.
+                // So we want to report it as an ApplicationError to fail the request and not process more events.
+                try
                 {
-                    var task = entry.Key.Invoke(entry.Value);
-                    if (!task.IsCompletedSuccessfully)
+                    while (events.TryPop(out var entry))
                     {
-                        return FireOnStartingAwaited(task, onStarting);
+                        await entry.Key.Invoke(entry.Value);
                     }
                 }
-            }
-            catch (Exception ex)
-            {
-                ReportApplicationError(ex);
-            }
-
-            return Task.CompletedTask;
-        }
-
-        private async Task FireOnStartingAwaited(Task currentTask, Stack<KeyValuePair<Func<object, Task>, object>> onStarting)
-        {
-            try
-            {
-                await currentTask;
-
-                while (onStarting.TryPop(out var entry))
+                catch (Exception ex)
                 {
-                    await entry.Key.Invoke(entry.Value);
+                    protocol.ReportApplicationError(ex);
                 }
             }
-            catch (Exception ex)
-            {
-                ReportApplicationError(ex);
-            }
         }
 
         protected Task FireOnCompleted()
         {
             var onCompleted = _onCompleted;
-
-            if (onCompleted == null || onCompleted.Count == 0)
+            if (onCompleted?.Count > 0)
             {
-                return Task.CompletedTask;
+                return ProcessEvents(this, onCompleted);
```

Comment:
If one of the other onCompleted callbacks throws, do we still continue looping over the other callbacks and executing? Unlike the onStarting callbacks, it's important that every onCompleted callback is invoked. I thought we had a test for this.

Reply:
Nvm. I just saw this was calling a local function.